api/src/main/java/org/openmrs/module/locationbasedaccess/PatientSearchAdvise.java

```diff
@@ -0,0 +1,94 @@
+import org.springframework.aop.support.StaticMethodMatcherPointcutAdvisor;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * It will be used to implement the AOP Methods for PatientServiceImpl class
+ */
+public class PatientSearchAdvise extends StaticMethodMatcherPointcutAdvisor implements Advisor {
+
+    protected static final Log log = LogFactory.getLog(PatientSearchAdvise.class);
+    public static final  String LOCATION_ATTRIBUTE_GLOBAL_PROPERTY_NAME = "locationbasedaccess.locationAttributeUuid";
+
+    @Override
+    public boolean matches(Method method, Class targetClass) {
+        if (method.getName().equals("getPatients")) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Advice getAdvice() {
+        return new PatientSearchControlAdvice();
+    }
+
+    /**
+     * This inner class will be used to intercept the caught method using AOP
+     */
+    private class PatientSearchControlAdvice implements MethodInterceptor {
+
+        public Object invoke(MethodInvocation invocation) throws Throwable {
+            Integer sessionLocationId = Context.getUserContext().getLocationId();
+            if(sessionLocationId != null) {
+                Location sessionLocation = Context.getLocationService().getLocation(sessionLocationId);
+                List<Patient> patientList = (List<Patient>)invocation.proceed();
+                for (Iterator<Patient> iterator = patientList.iterator(); iterator.hasNext(); ) {
+                    Patient patient  = iterator.next();
+                    if(!isPatientInLocation(patient, sessionLocation)) {
+                        iterator.remove();
+                    }
+                }
+                return patientList;
+            }
+            else {
+                // If the sessionLocationId is null, then return a empty list
+                log.debug("Search Patient : Null Session Location in the UserContext");
+                return new ArrayList<Patient>();
+            }
+        }
+
+        public Boolean isPatientInLocation(Patient patient, Location sessionLocation) {
+            String locationAttributeUuid = Context.getAdministrationService().getGlobalProperty(LOCATION_ATTRIBUTE_GLOBAL_PROPERTY_NAME);
+            if (StringUtils.isNotBlank(locationAttributeUuid)) {
+                final PersonAttribute personAttribute = patient.getAttribute(Context.getPersonService().getPersonAttributeTypeByUuid(locationAttributeUuid));
+                if (personAttribute == null) {
+                    log.debug("Patient AccessLocation attribute missing for Patient - " + patient.getGivenName());
+                }
+                else {
+                    if(!personAttribute.getValue().equals(sessionLocation.getUuid())) {
+                        return false;
+                    }
+                }
+            }
+            return true;
```

Comment:
According to your code, what are the cases when true will be returned?

Reply:
Following are the cases when true will be returned,
1. If the system doesn't have the required global property to track the personAttributeType (like ignoring this filter) 
2. If the patient doesn't have the locationAttribute to check the location (probably for the existing patients) 
3. If the patient has the locationAttribute and it's equal to the sessionLocation of the logged in user. 

Only one case we have now to return FALSE, 
1. If the patient has the locationAttribute and it's ##NOT## equal to the sessionLocation of the logged in user.

Reply:
Let us use a safer default. If a patient is not having the selected location, they should not appear at all.

Reply:
Applied @dkayiwa

Reply:
You have not applied changes to any of my comments on this method.

Reply:
Your current code has not addressed my comments yet. Slow down, and give it a second look. As yourself, "Why is Daniel insisting? Could i have missed something"?